lib/rate_map.c

```diff
@@ -235,66 +235,124 @@ rate_map_shift_by_mass(rate_map_t *self, double pos, double mass)
+        if (!(array[idx - 1] <= array[idx])) {
+            return false;
+        }
+    }
+    return true;
+}
+
+static bool
+fast_search_lookup_valid(fast_search_lookup_t *self)
+{
+    const double *start, *stop;
+
+    if (!(self->query_multiplier >= 0.0)) { // NaN not valid
+        return false;
+    }
+    if (self->num_lookups < 2) {
+        return false;
+    }
+    start = self->lookups[0];
+    stop = self->lookups[self->num_lookups - 1];
+    if (stop <= start || start[0] != 0.0) {
+        return false;
+    }
+    return valid_sorted_nonempty_array(start, (size_t)(stop - start));
+}
+
 static int
 fast_search_lookup_init_lookups(
-    fast_search_lookup_t *self, const double *values, size_t n_values)
+    fast_search_lookup_t *self, const double *elements, size_t n_elements)
 {
     int ret = 0;
-    const double *lower, *strict_upper;
+    const double *ptr, *stop;
     uint32_t i;
-    double query = -INFINITY;
+    double query = 0.0;
 
-    lower = values;
+    ptr = elements;
+    stop = elements + n_elements;
     for (i = 0; i < self->num_lookups; i++) {
-        while (*lower <= query) {
-            lower++;
-        }
-        self->lookups[i].start = lower;
-        strict_upper = lower;
-        query = ldexp(i, -(self->power_shift));
-        assert(strict_upper < values + n_values);
-        while (strict_upper < values + n_values && *strict_upper <= query) {
-            strict_upper++;
+        while (ptr < stop && *ptr < query) {
+            ptr++;
         }
-        const int64_t num = strict_upper - lower;
-        if (num > INT32_MAX) {
-            ret = MSP_ERR_BAD_PARAM_VALUE; // TODO: consider new err code
-            goto out;
+        self->lookups[i] = ptr;
+        if (ptr < stop) {
+            query = (i + 1) / self->query_multiplier;
         }
-        self->lookups[i].num = (size_t) num;
+    }
+    if (!fast_search_lookup_valid(self)) {
+        ret = MSP_ERR_ASSERTION_FAILED;
+        goto out;
     }
 out:
     return ret;
 }
 
-static inline long
-ceil_to_long(double x)
+/* Returns the least power of 2 higher than (or equal to) `x` for positive numbers,
+ * otherwise returns zero.
+ */
+static double
+higher_power_of_2(double x)
 {
-    return (long) ceil(x); // TODO: test use of fast non-portable equivalent
+    assert(x >= 0);
+    return (x > 0 ? exp2(ceil(logb(x))) : 0.0);
 }
 
+/* PRE-CONDITIONS:
+ *     1) `elements` must point to array of doubles starting at exactly `0.0`
+ *     2) `elements` must be non-empty non-decreasing and with no NaN
+ */
 int
 fast_search_lookup_alloc(
-    fast_search_lookup_t *self, const double *values, size_t n_values)
+    fast_search_lookup_t *self, const double *elements, size_t n_elements)
 {
     int ret = 0;
-    const double max = values[n_values - 1];
-    const int power = ilogb((double) n_values) + 1;
+    double max_element, upper_bound;
+
+    const size_t max_input_size = 1L << (DBL_MANT_DIG - 1); // 4096 terabytes
```

Comment:
Is the literal ``1L`` safe here? I've had issues before with 1L being interpreted as a 4 byte integer. Safer use ``1LL``?

Reply:
Good point. Making sure it's not compiled as 32-bits with 1LL does seem safer.

Reply:
Good eye @jeromekelleher thx! This fixed a unit test failure on appveyor running on Windows. After saw it fail on (I suspect) Windows more than once I was thinking "oh boy, this is going to be a pain in the A to figure out." So you saved me a lot of time. 64-bit Windows keeps long as 32-bit.

Reply:
Does this have the intended behaviour on a 32 bit platform? For 32 bit `size_t`, this will evaluate to `0`, right?  (Does msprime even try to support 32 bit platforms?)

Reply:
Good catch @grahamgower! Indeed it should be a `1LL` shifted and not `1L`. It was causing Windows unit tests to fail. `long` is 32-bit on 64-bit Windows.

Reply:
To be clear, I meant that on a 32 bit platform, `size_t max_input_size` will be a 32 bit variable. So,  `1LL << (53 - 1)` will be cast to a 32 bit variable. So on a 32 bit platform you'll have `max_input_size = 0` even if you change the right hand side from 1L to 1LL.

Reply:
Oh, now I follow you. That makes sense. I find the out-of-date diffs that github shows in the conversation tab a bit confusing. This has got to be something like the fifth time that I have misunderstood comments because my brain seems to pay more attention C code than English. :)

I doubt anybody is using 32-bit and I suspect 32-bits can be dropped from support. I'll add an #ifdef #error so that it's clear. @jeromekelleher knows whether this blatant disrespect for 32-bit platforms is acceptable or not.